core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -867,6 +869,26 @@ class Partition(val topicPartition: TopicPartition,
     }
   }
 
+  private def createLogInAssignedDirectoryId(partitionState: LeaderAndIsrPartitionState, highWatermarkCheckpoints: OffsetCheckpoints, topicId: Option[Uuid], targetLogDirectoryId: Option[Uuid]): Unit = {
+    targetLogDirectoryId match {
+      case Some(directoryId) =>
+        if (logManager.onlineLogDirId(directoryId)) {
+          if (logManager.getLog(topicPartition).isEmpty) {
+            createLogIfNotExists(partitionState.isNew, isFutureReplica = false, highWatermarkCheckpoints, topicId, targetLogDirectoryId)
+          } else {
+            logger.info(s"Topic partition ${topicPartition.topic()}-${topicPartition.partition()} already exists.")
+          }
+        } else if ((!logManager.onlineLogDirId(directoryId) && !logManager.hasOfflineLogDirs()) || directoryId == DirectoryId.UNASSIGNED) {
+          // create on next selected directory if the directory Id is LOST or unknown and no offline directories or unassigned
+          createLogIfNotExists(partitionState.isNew, isFutureReplica = false, highWatermarkCheckpoints, topicId, None)
```

Comment:
The partition may exist in some other log dir

Reply:
I think if we have a directory ID defined, we can safely set `isNew` to `directoryId == DirectoryId.UNASSIGNED`. And then I think we don't need the check above for offline directories.

Reply:
Good call, updated the pr to reflect this.